packages/common/src/index-api.ts

```diff
@@ -24,13 +24,76 @@ export type BackwardPagination = {
  */
 export type Pagination = ForwardPagination | BackwardPagination
 
+/**
+ * Boolean field value filter
+ */
+export type BooleanValueFilter = { isNull: boolean } | { equalTo: boolean }
+
+/**
+ * Enum field value filter
+ */
+export type EnumValueFilter =
+  | { isNull: boolean }
+  | { equalTo: string }
+  | { notEqualTo: string }
+  | { in: Array<string> }
+  | { notIn: Array<string> }
+
+/**
+ * String or number field value filter
+ */
+export type ScalarValueFilter<T extends string | number = string | number> =
+  | { isNull: boolean }
+  | { equalTo: T }
+  | { notEqualTo: T }
+  | { in: Array<T> }
+  | { notIn: Array<T> }
+  | { lessThan: T }
+  | { lessThanOrEqualTo: T }
+  | { greaterThan: T }
+  | { greaterThanOrEqualTo: T }
+
+/**
+ * Any supported field value filter on an object
+ */
+export type AnyValueFilter = BooleanValueFilter | EnumValueFilter | ScalarValueFilter
+
+/**
+ * Mapping of object keys to value filter
+ */
+export type ObjectFilter = Record<string, AnyValueFilter>
+
+/**
+ * Advanced query filters on a document fields
+ */
+export type QueryFilters =
+  | { doc: ObjectFilter }
```

Comment:
can you explain more how `doc` filters would work?

Reply:
I think they'd work by iterating over all the keys and matching value filters to create the SQL query, something like the following:

```ts
// Build a single condition filter string
function buildConditionFilter(condition: keyof AnyValueFilter, value: string): string {
  switch (condition) {
    case 'equalTo':
      return `= ${value}`
    case 'notEqualTo':
      return `<> ${value}`
    case 'in':
      return `IN (${value})`
    // ... other cases covering all supported conditions
  }
}

const RANGE_GREATER_CONDITIONS = ['greaterThan', 'greaterThanOrEqualTo']
const RANGE_LOWER_CONDITIONS = ['lowerThan', 'lowerThanOrEqualTo']

// Build the filter string for a single field
function buildValueFilter(field: string, filter: AnyValueFilter): string {
  const conditions = Object.keys(filter)
  // Check valid number of conditions
  if (conditions.length === 0 || conditions.length > 2) {
    throw new Error(`Invalid number of conditions in filter for field ${field}`)
  }

  if (conditions.length === 1) {
    const condition = conditions[0]
    const value = Array.isArray(filter[condition]) ? filter[condition].join(', ') : filter[condition].toString()
    return `${field} ${buildConditionFilter(condition, value)}`
  }

  // Check valid range when 2 conditions are provided
  const [conditionA, conditionB] = conditions
  if (RANGE_GREATER_CONDITIONS.includes(conditionA) && RANGE_LOWER_CONDITIONS.includes(conditionB) || RANGE_GREATER_CONDITIONS.includes(conditionB) && RANGE_LOWER_CONDITIONS.includes(conditionA)) {
    const filterA = buildConditionFilter(conditionA, filters[conditionA].toString())
    const filterB = buildConditionFilter(conditionB, filters[conditionB].toString())
    return `${field} ${filterA} AND ${field} ${filterB}`
  }
  throw new Error(`Invalid range filter for field ${field}`)
}

// Build the filter string for all fields in an object
function buildObjectFilter(objectFilter: ObjectFilter): string {
  const parts = Object.entries(objectFilter).map(([field, filter]) => buildValueFilter(field, filter))
  return parts.join(' AND ')
}
```

In our case we'll likely want to leverage Knex to build the queries rather than generating SQL directly as I did here, but the logic should be similar I think.

Reply:
okay, so fields that are specified will have filters set, but fields not included will not be filtered on.  If you wanted to specify that the doc matched exactly a given set of fields *and no others*, I guess the only way to do that would be to know all possible optional fields in the schema and have a filter to explicitly check that they are `undefined`?  Would that even work?  Can you send `undefined` as a value for a filter?

Reply:
Maybe instead of `doc` we can call this something else (`fields`, `filter`, `match`, `matchFields`, etc).  Then if in the future we wanted an "exact-match" operator for a full document structure we could add `doc` later.

Reply:
I think I like `matchFields` most?

Or do we need to call this `doc` to maintain compatibility with Graphile?

Reply:
Actually, if I'm reading this right, I think Graphile calls this `filter`?  Which I'd also be fine with.
https://github.com/graphile-contrib/postgraphile-plugin-connection-filter/blob/master/docs/examples.md

Reply:
Yes we can use a different name than `doc` if you think it's confusing, the main thing is that we need a specific key to differentiate from the `and`, `or` and `not` operators.
Graphile uses `filter` as the input name, same as what we have as `queryFilters` in the `BaseQuery` objects so I'd rather use a different term so we don't have `queryFilters.filter`. `fields` sounds good to me, `where` could also be a possibility I think. `match` could be misleading if we ever supported regular expressions, but could be another option.

Reply:
`fields` or `where` both sound good to me

Reply:
I think I've seen `where` used more often than I've seen `fields`, so I guess I very very slightly lean that way?  But yeah its very slight, I think `fields` would be totally fine also